File: bindings/python/iota_sdk/types/output.py
# Copyright 2023 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations
from enum import IntEnum
from typing import Dict, Optional, List, TypeAlias, Union, Any
from dataclasses import dataclass, field
from dataclasses_json import config
from iota_sdk.types.common import HexStr, json, EpochIndex
from iota_sdk.types.feature import deserialize_features, SenderFeature, IssuerFeature, MetadataFeature, TagFeature, NativeTokenFeature
from iota_sdk.types.token_scheme import SimpleTokenScheme
from iota_sdk.types.unlock_condition import deserialize_unlock_conditions, AddressUnlockCondition, StateControllerAddressUnlockCondition, GovernorAddressUnlockCondition, StorageDepositReturnUnlockCondition, TimelockUnlockCondition, ExpirationUnlockCondition, ImmutableAccountAddressUnlockCondition
from iota_sdk.types.address import AccountAddress


class OutputType(IntEnum):
    """Output types.

    Attributes:
        Basic (0): A basic output.
        Account (1): An account output.
        Anchor (2): An anchor output.
        Foundry (3): A foundry output.
        Nft (4): An NFT output.
        Delegation (5): A delegation output.

    """
    Basic = 0
    Account = 1
    Anchor = 2
    Foundry = 3
    Nft = 4
    Delegation = 5


@json
@dataclass
class BasicOutput:
    """Describes a basic output.

    Attributes:
        amount :
            The base coin amount of the output.
        mana :
            Amount of stored Mana held by this output.
        unlock_conditions :
            The conditions to unlock the output.
        features :
            Features that add utility to the output but do not impose unlocking conditions.
        type :
            The type of output.
    """
    type: int = field(
        default_factory=lambda: int(
            OutputType.Basic),
        init=False)
    amount: int = field(metadata=config(
        encoder=str
    ))
    mana: int = field(metadata=config(
        encoder=str
    ))
    unlock_conditions: List[Union[AddressUnlockCondition, ExpirationUnlockCondition, StorageDepositReturnUnlockCondition,
                                  TimelockUnlockCondition]] = field(metadata=config(
                                                                    decoder=deserialize_unlock_conditions
                                                                    ))
    features: Optional[List[Union[SenderFeature,
                            MetadataFeature, TagFeature, NativeTokenFeature]]] = field(default=None,
                                                                                       metadata=config(
                                                                                           decoder=deserialize_features
                                                                                       ))


@json
@dataclass
class AccountOutput:
    """Describes an account output.

    Attributes:
        amount :
            The base coin amount of the output.
        mana :
            Amount of stored Mana held by this output.
        unlock_conditions:
            The conditions to unlock the output.
        account_id :
            The account ID if it's an account output.
        foundry_counter :
            A counter that denotes the number of foundries created by this account output.
        features :
            Features that add utility to the output but do not impose unlocking conditions.
        immutable_features :
            Features that add utility to the output but do not impose unlocking conditions. These features need to be kept in future transitions of the UTXO state machine.
        type :
            The type of output.
    """
    type: int = field(
        default_factory=lambda: int(
            OutputType.Account),
        init=False)
    amount: int = field(metadata=config(
        encoder=str
    ))
    mana: int = field(metadata=config(
        encoder=str
    ))
    account_id: HexStr
    foundry_counter: int
    unlock_conditions: List[AddressUnlockCondition] = field(
        metadata=config(
            decoder=deserialize_unlock_conditions
        ))
    features: Optional[List[Union[SenderFeature,
                            MetadataFeature]]] = field(default=None,
                                                       metadata=config(
                                                           decoder=deserialize_features
                                                       ))
    immutable_features: Optional[List[Union[IssuerFeature,
                                            MetadataFeature]]] = field(default=None,
                                                                       metadata=config(
                                                                           decoder=deserialize_features
                                                                       ))


@json
@dataclass
class AnchorOutput:
    """Describes an anchor output.

    Attributes:
        amount :
            The base coin amount of the output.
        mana :
            Amount of stored Mana held by this output.
        anchor_id :
            The anchor ID if it's an anchor output.
        state_index :
            A counter that must increase by 1 every time the anchor is state transitioned.
        unlock_conditions:
            The conditions to unlock the output.
        features :
            Features that add utility to the output but do not impose unlocking conditions.
        immutable_features :
            Features that add utility to the output but do not impose unlocking conditions. These features need to be kept in future transitions of the UTXO state machine.
        type :
            The type of output.
    """
    type: int = field(
        default_factory=lambda: int(
            OutputType.Anchor),
        init=False)
    amount: int = field(metadata=config(
        encoder=str
    ))
    mana: int = field(metadata=config(
        encoder=str
    ))
    anchor_id: HexStr
    state_index: int
    unlock_conditions: List[Union[StateControllerAddressUnlockCondition,
                                  GovernorAddressUnlockCondition]] = field(
        metadata=config(
            decoder=deserialize_unlock_conditions
        ))
    features: Optional[List[Union[SenderFeature,
                            MetadataFeature]]] = field(default=None,
                                                       metadata=config(
                                                           decoder=deserialize_features
                                                       ))
    immutable_features: Optional[List[Union[IssuerFeature,
                                            MetadataFeature]]] = field(default=None,
                                                                       metadata=config(
                                                                           decoder=deserialize_features
                                                                       ))


@json
@dataclass
class FoundryOutput:
    """Describes a foundry output.

    Attributes:
        amount :
            The base coin amount of the output.
        unlock_conditions :
            The conditions to unlock the output.
        features :
            Features that add utility to the output but do not impose unlocking conditions.
        immutable_features :
            Features that add utility to the output but do not impose unlocking conditions. These features need to be kept in future transitions of the UTXO state machine.
        serial_number :
            The serial number of the foundry with respect to the controlling account.
        token_scheme :
            Defines the supply control scheme of the tokens controlled by the foundry. Currently only a simple scheme is supported.
        type :
            The type of output.
    """
    type: int = field(
        default_factory=lambda: int(
            OutputType.Foundry),
        init=False)
    amount: int = field(metadata=config(
        encoder=str
    ))
    serial_number: int
    token_scheme: SimpleTokenScheme
    unlock_conditions: List[ImmutableAccountAddressUnlockCondition]
    features: Optional[List[Union[MetadataFeature, NativeTokenFeature]]] = field(default=None,
                                                                                 metadata=config(
                                                                                     decoder=deserialize_features
                                                                                 ))
    immutable_features: Optional[List[MetadataFeature]] = field(default=None,
                                                                metadata=config(
                                                                    decoder=deserialize_features
                                                                ))


@json
@dataclass
class NftOutput:
    """Describes an NFT output.

    Attributes:
        amount :
            The base coin amount of the output.
        mana :
            Amount of stored Mana held by this output.
        unlock_conditions :
            The conditions to unlock the output.
        nft_id :
            The NFT ID if it's an NFT output.
        features :
            Features that add utility to the output but do not impose unlocking conditions.
        immutable_features :
            Features that add utility to the output but do not impose unlocking conditions. These features need to be kept in future transitions of the UTXO state machine.
        type :
            The type of output.
    """
    type: int = field(default_factory=lambda: int(OutputType.Nft), init=False)
    amount: int = field(metadata=config(
        encoder=str
    ))
    mana: int = field(metadata=config(
        encoder=str
    ))
    nft_id: HexStr
    unlock_conditions: List[Union[AddressUnlockCondition, ExpirationUnlockCondition,
                                  StorageDepositReturnUnlockCondition, TimelockUnlockCondition]] = field(
        metadata=config(
            decoder=deserialize_unlock_conditions
        ))
    features: Optional[List[Union[SenderFeature,
                            MetadataFeature, TagFeature]]] = field(default=None,
                                                                   metadata=config(
                                                                       decoder=deserialize_features
                                                                   ))
    immutable_features: Optional[List[Union[
        IssuerFeature, MetadataFeature]]] = field(default=None,
                                                  metadata=config(
                                                      decoder=deserialize_features
                                                  ))


@json
@dataclass
class DelegationOutput:
    """An output which delegates its contained IOTA coins as voting power to a validator.

    Attributes:
        amount: The amount of IOTA coins held by the output.
        delegated_amount: The amount of delegated IOTA coins.
        delegation_id: Unique identifier of the Delegation Output
        validator_address: The Account Address of the validator to which this output is delegating.
        start_epoch: The index of the first epoch for which this output delegates.
        end_epoch: The index of the last epoch for which this output delegates.
        unlock_conditions: Define how the output can be unlocked in a transaction.
        type: The type of output.
    """
    type: int = field(default_factory=lambda: int(
        OutputType.Delegation), init=False)
    amount: int = field(metadata=config(
        encoder=str
    ))
    delegated_amount: int = field(metadata=config(
        encoder=str
    ))
    delegation_id: HexStr
    validator_address: AccountAddress
    start_epoch: EpochIndex
    end_epoch: EpochIndex
    unlock_conditions: List[AddressUnlockCondition] = field(metadata=config(
        decoder=deserialize_unlock_conditions
    ))


Output: TypeAlias = Union[BasicOutput,
                          AccountOutput,
                          AnchorOutput,
                          FoundryOutput,
                          NftOutput,
                          DelegationOutput]


def deserialize_output(d: Dict[str, Any]) -> Output:
    """
    Takes a dictionary as input and returns an instance of a specific class based on the value of the 'type' key in the dictionary.

    Arguments:
    * `d`: A dictionary that is expected to have a key called 'type' which specifies the type of the returned value.
    """
    output_type = d['type']
    if output_type == OutputType.Basic:
        return BasicOutput.from_dict(d)
    if output_type == OutputType.Account:
        return AccountOutput.from_dict(d)
    if output_type == OutputType.Anchor:
        return AnchorOutput.from_dict(d)
    if output_type == OutputType.Foundry:
        return FoundryOutput.from_dict(d)
    if output_type == OutputType.Nft:
        return NftOutput.from_dict(d)
    if output_type == OutputType.Delegation:
        return DelegationOutput.from_dict(d)
    raise Exception(f'invalid output type: {output_type}')


def deserialize_outputs(dicts: List[Dict[str, Any]]) -> List[Output]:
    """
    Takes a list of dictionaries as input and returns a list with specific instances of classes based on the value of the 'type' key in the dictionary.

    Arguments:
    * `dicts`: A list of dictionaries that are expected to have a key called 'type' which specifies the type of the returned value.
    """
    return list(map(deserialize_output, dicts))

File: bindings/nodejs/lib/types/block/output/output.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
    UnlockCondition,
    UnlockConditionDiscriminator,
} from './unlock-condition';
import { Feature, FeatureDiscriminator, NativeTokenFeature } from './feature';

// Temp solution for not double parsing JSON
import { plainToInstance, Type } from 'class-transformer';
import { HexEncodedString, NumericString, u64 } from '../../utils';
import { TokenScheme, TokenSchemeDiscriminator } from './token-scheme';
import { AccountId, NftId, AnchorId, DelegationId } from '../id';
import { EpochIndex } from '../../block/slot';
import { NativeToken } from '../../models/native-token';
import { Address, AddressDiscriminator, AccountAddress } from '../address';

export type OutputId = HexEncodedString;

/**
 * All of the output types.
 */
enum OutputType {
    /** A Basic output. */
    Basic = 0,
    /** An Account output. */
    Account = 1,
    /** An Anchor output. */
    Anchor = 2,
    /** A Foundry output. */
    Foundry = 3,
    /** An NFT output. */
    Nft = 4,
    /** A Delegation output. */
    Delegation = 5,
}

/**
 * The base class for outputs.
 */
abstract class Output {
    readonly amount: NumericString;

    /**
     * The type of output.
     */
    readonly type: OutputType;

    /**
     * @param type The type of output.
     * @param amount The amount of the output as big-integer or string.
     */
    constructor(type: OutputType, amount: u64 | NumericString) {
        this.type = type;
        if (typeof amount == 'bigint') {
            this.amount = amount.toString(10);
        } else {
            this.amount = amount;
        }
    }

    /**
     * Get the amount of the output.
     */
    getAmount(): u64 {
        return BigInt(this.amount);
    }

    /**
     * Parse an output from a plain JS JSON object.
     */
    public static parse(data: any): Output {
        if (data.type == OutputType.Basic) {
            return plainToInstance(BasicOutput, data) as any as BasicOutput;
        } else if (data.type == OutputType.Account) {
            return plainToInstance(AccountOutput, data) as any as AccountOutput;
        } else if (data.type == OutputType.Anchor) {
            return plainToInstance(AnchorOutput, data) as any as AnchorOutput;
        } else if (data.type == OutputType.Foundry) {
            return plainToInstance(FoundryOutput, data) as any as FoundryOutput;
        } else if (data.type == OutputType.Nft) {
            return plainToInstance(NftOutput, data) as any as NftOutput;
        } else if (data.type == OutputType.Delegation) {
            return plainToInstance(
                DelegationOutput,
                data,
            ) as any as DelegationOutput;
        }
        throw new Error('Invalid JSON');
    }
}

/**
 * The base class for common outputs.
 */
abstract class CommonOutput extends Output {
    /**
     * The unlock conditions for the output.
     */
    @Type(() => UnlockCondition, {
        discriminator: UnlockConditionDiscriminator,
    })
    readonly unlockConditions: UnlockCondition[];

    /**
     * The features contained by the output.
     */
    @Type(() => Feature, {
        discriminator: FeatureDiscriminator,
    })
    readonly features?: Feature[];

    /**
     * @param type The type of output.
     * @param amount The amount of the output.
     * @param unlockConditions Conditions to unlock the output.
     */
    constructor(
        type: OutputType,
        amount: u64,
        unlockConditions: UnlockCondition[],
    ) {
        super(type, amount);
        this.unlockConditions = unlockConditions;
    }

    /**
     * The native token held by the output.
     */
    getNativeToken(): NativeToken | undefined {
        if (!this.features) {
            return undefined;
        }

        for (const feature of this.features) {
            if (feature instanceof NativeTokenFeature) {
                return (feature as NativeTokenFeature).asNativeToken();
            }
        }
        return undefined;
    }
}

/**
 * A Basic output.
 */
class BasicOutput extends CommonOutput {
    /**
     * The amount of (stored) Mana held by the output.
     */
    readonly mana: u64;

    /**
     * @param amount The amount of the output.
     * @param unlockConditions The unlock conditions for the output.
     */
    constructor(amount: u64, mana: u64, unlockConditions: UnlockCondition[]) {
        super(OutputType.Basic, amount, unlockConditions);
        this.mana = mana;
    }
}

/**
 * Base class for immutable feature outputs.
 */
abstract class ImmutableFeaturesOutput extends CommonOutput {
    @Type(() => Feature, {
        discriminator: FeatureDiscriminator,
    })
    readonly immutableFeatures?: Feature[];

    /**
     * @param type The type of output.
     * @param amount The amount of the output.
     * @param unlockConditions The unlock conditions for the output.
     */
    constructor(
        type: OutputType,
        amount: u64,
        unlockConditions: UnlockCondition[],
    ) {
        super(type, amount, unlockConditions);
    }
}

/**
 * An Account output.
 */
class AccountOutput extends ImmutableFeaturesOutput {
    /**
     * Unique identifier of the account, which is the BLAKE2b-256 hash of the Output ID that created it.
     * Unless its a newly created account, then the id is zeroed.
     */
    readonly accountId: AccountId;
    /**
     * A counter that denotes the number of foundries created by this account output.
     */
    readonly foundryCounter: number;
    /**
     * The amount of (stored) Mana held by the output.
     */
    readonly mana: u64;

    /**
     * @param amount The amount of the output.
     * @param mana The amount of stored mana.
     * @param accountId The account ID as hex-encoded string.
     * @param foundryCounter A counter that denotes the number of foundries created by this account output.
     * @param unlockConditions The unlock conditions of the output.
     */
    constructor(
        amount: u64,
        mana: u64,
        accountId: AccountId,
        foundryCounter: number,
        unlockConditions: UnlockCondition[],
    ) {
        super(OutputType.Account, amount, unlockConditions);
        this.accountId = accountId;
        this.foundryCounter = foundryCounter;
        this.mana = mana;
    }
}

/**
 * An Anchor output.
 */
class AnchorOutput extends ImmutableFeaturesOutput {
    /**
     * Unique identifier of the anchor, which is the BLAKE2b-256 hash of the Output ID that created it.
     * Unless its a newly created anchor, then the id is zeroed.
     */
    readonly anchorId: AnchorId;
    /**
     * A counter that must increase by 1 every time the anchor output is state transitioned.
     */
    readonly stateIndex: number;
    /**
     * The amount of (stored) Mana held by the output.
     */
    readonly mana: u64;

    /**
     * @param amount The amount of the output.
     * @param mana The amount of stored mana.
     * @param anchorId The anchor ID as hex-encoded string.
     * @param stateIndex A counter that must increase by 1 every time the anchor output is state transitioned.
     * @param unlockConditions The unlock conditions of the output.
     */
    constructor(
        amount: u64,
        mana: u64,
        anchorId: AnchorId,
        stateIndex: number,
        unlockConditions: UnlockCondition[],
    ) {
        super(OutputType.Account, amount, unlockConditions);
        this.anchorId = anchorId;
        this.stateIndex = stateIndex;
        this.mana = mana;
    }
}

/**
 * A Foundry output.
 */
class FoundryOutput extends ImmutableFeaturesOutput {
    /**
     * The serial number of the Foundry with respect to the controlling account.
     */
    readonly serialNumber: number;

    /**
     * The token scheme for the Foundry.
     */
    @Type(() => TokenScheme, {
        discriminator: TokenSchemeDiscriminator,
    })
    readonly tokenScheme: TokenScheme;

    /**
     * @param amount The amount of the output.
     * @param serialNumber The serial number of the Foundry with respect to the controlling account.
     * @param unlockConditions The unlock conditions of the output.
     * @param tokenScheme The token scheme for the Foundry.
     */
    constructor(
        amount: u64,
        serialNumber: number,
        unlockConditions: UnlockCondition[],
        tokenScheme: TokenScheme,
    ) {
        super(OutputType.Foundry, amount, unlockConditions);
        this.serialNumber = serialNumber;
        this.tokenScheme = tokenScheme;
    }
}

/**
 * An NFT output.
 */
class NftOutput extends ImmutableFeaturesOutput {
    /**
     * Unique identifier of the NFT, which is the BLAKE2b-256 hash of the Output ID that created it.
     * Unless its newly minted, then the id is zeroed.
     */
    readonly nftId: NftId;

    /**
     * The amount of (stored) Mana held by the output.
     */
    readonly mana: u64;

    /**
     * @param amount The amount of the output.
     * @param mana The amount of stored mana.
     * @param nftId The NFT ID as hex-encoded string.
     * @param unlockConditions The unlock conditions of the output.
     */
    constructor(
        amount: u64,
        mana: u64,
        nftId: NftId,
        unlockConditions: UnlockCondition[],
    ) {
        super(OutputType.Nft, amount, unlockConditions);
        this.nftId = nftId;
        this.mana = mana;
    }
}

/**
 * A Delegation output.
 */
class DelegationOutput extends Output {
    /**
     * The amount of delegated coins.
     */
    readonly delegatedAmount: u64;
    /**
     * Unique identifier of the Delegation Output, which is the BLAKE2b-256 hash of the Output ID that created it.
     */
    readonly delegationId: DelegationId;
    /**
     * The Account ID of the validator to which this output is delegating.
     */
    @Type(() => Address, {
        discriminator: AddressDiscriminator,
    })
    readonly validatorAddress: Address;
    /**
     * The index of the first epoch for which this output delegates.
     */
    readonly startEpoch: EpochIndex;
    /**
     * The index of the last epoch for which this output delegates.
     */
    readonly endEpoch: EpochIndex;
    /**
     * The unlock conditions for the output.
     */
    @Type(() => UnlockCondition, {
        discriminator: UnlockConditionDiscriminator,
    })
    readonly unlockConditions: UnlockCondition[];

    /**
     * @param amount The amount of the output.
     * @param delegatedAmount The amount of delegated coins.
     * @param delegationId Unique identifier of the Delegation Output, which is the BLAKE2b-256 hash of the Output ID that created it.
     * @param validatorAddress The Account address of the validator to which this output is delegating.
     * @param startEpoch The index of the first epoch for which this output delegates.
     * @param endEpoch The index of the last epoch for which this output delegates.
     * @param unlockConditions The unlock conditions of the output.
     */
    constructor(
        amount: u64,
        delegatedAmount: u64,
        delegationId: DelegationId,
        validatorAddress: AccountAddress,
        startEpoch: EpochIndex,
        endEpoch: EpochIndex,
        unlockConditions: UnlockCondition[],
    ) {
        super(OutputType.Delegation, amount);
        this.delegatedAmount = delegatedAmount;
        this.delegationId = delegationId;
        this.validatorAddress = validatorAddress;
        this.startEpoch = startEpoch;
        this.endEpoch = endEpoch;
        this.unlockConditions = unlockConditions;
    }
}

const OutputDiscriminator = {
    property: 'type',
    subTypes: [
        { value: BasicOutput, name: OutputType.Basic as any },
        { value: AccountOutput, name: OutputType.Account as any },
        { value: AnchorOutput, name: OutputType.Anchor as any },
        { value: FoundryOutput, name: OutputType.Foundry as any },
        { value: NftOutput, name: OutputType.Nft as any },
        { value: DelegationOutput, name: OutputType.Delegation as any },
    ],
};

export {
    OutputDiscriminator,
    Output,
    OutputType,
    CommonOutput,
    BasicOutput,
    AccountOutput,
    AnchorOutput,
    FoundryOutput,
    NftOutput,
    DelegationOutput,
};
